doc/mgmt/OpenConfig_Path_Translation_Root_Expansion_HLD.md

```diff
@@ -0,0 +1,338 @@
+  ...
+  translCtx := pathtransl.TranslSetRequest(req)
+  ... // Data clients process req, and return SetResponse resp
+  pathtransl.TranslSetResponse(resp, translCtx)
+  return resp, err
+}
+```
+
+##### SubscribeRequest
+Translation of SubscribeRequest happens inside the Subscribe Client’s Run process before request is passed to transclient.
+
+(gnmi_server/client_subscribe.go)
+```
+func (c *Client) Run(stream gnmipb.GNMI_SubscribeServer) (err error) {
+  ...
+  c.translCtx := pathtransl.TranslSubscribeRequest(req)
+  ...
+}
+```
+
+##### SubscribeResponse
+Translation of SubscribeRequest happens inside the Subscribe Client’s Send process before notification response is sent.
+
+(gnmi_server/client_subscribe.go)
+```
+func (c *Client) send(stream gnmipb.GNMI_SubscribeServer) error {
+  ...
+  pathtransl.TranslSubscribeResponse(resp, c.translCtx)
+  ...
+}
+```
+
+#### Single Root Translation
+Single Root means xpath `/openconfig`, which UMF currently does not support. We can use this translation layer to support single root as well.
+
+##### GetRequest
+Single Root GetRequest is translated into a GetRequest with all the supported OpenConfig paths as below:
+
+```
+Original GetRequest (single root)
+prefix:<origin:"openconfig">path:<>
+```
+
+```
+Translated GetRequest
+prefix:<>path:<elem:<name:"openconfig-interfaces:interfaces">>path:<elem:<name:"openconfig-lacp:lacp">>path:<elem:<name:"openconfig-platform:components">>path:<elem:<name:"openconfig-qos:qos">>path:<elem:<name:"openconfig-sampling:sampling">>path:<elem:<name:"openconfig-system:system">>
+```
+
+##### GetResponse
+Since the single root GetRequest is translated into multiple top-container paths, the JSON_IETF response will receive multiple notifications (one for each translated path). These multiple notifications need to be combined into one single JSON_IETF value and sent back to the client.
+Example:
+```
+Original GetResponse (with multiple notification)
+notification:<timestamp:1234566 prefix:<> update:<path:<elem:<name:"openconfig-interfaces:interfaces">>val:<json_ietf_val:"{\"openconfig-interfaces:interfaces\":{...}}“>
+update:<path:<elem:<name:"openconfig-platform:components">>val:<json_ietf_val:"{\"openconfig-platform:components\":{xxx}}">
+>
+```
+```
+Translated (combined) GetResponse
+notification:<timestamp:1234566
+prefix:<origin:"openconfig">
+update:<path:<> val:<json_ietf_val:
+"{
+  \"openconfig-interfaces:interfaces:\": {...},
+  \"openconfig-platform:components:\": {xxx}
+}“
+>>
+```
+
+##### SetRequest
+For a Single root SetRequest with JSON_IETF encoding, the JSON value is in the same format as in the translated GetResponse above.
+A mix of single root and non-single root SET is not supported. So for delete/update/replace, either there is only one root delete/update/replace, or non-single root ones.
+
+To translate the SetRequest:
+
+###### Single Root Delete
+It will be translated into multiple delete paths as show below:
```

Comment:
If I understand correctly, this will result in multiple CAS transactions in translib, right? If so, it's better to handle it with a single CAS (check and save, Redis MULTI) to reduce the chance of errors due to transactional failure